src/mscorlib/src/System/Text/EncodingForwarder.cs

```diff
@@ -192,6 +196,332 @@ public unsafe static int GetBytes(Encoding encoding, char* chars, int charCount,
+                    if (lengthEncoded < byteCount) {
+                        // Not all ASCII, use encoding's GetBytes for remaining conversion
+                        lengthEncoded += encoding.GetBytesFallback(input + lengthEncoded, charCount - lengthEncoded, output + lengthEncoded, byteCount - lengthEncoded, null);
+                    }
+                }
+            } else {
+                // Nothing to encode
+                lengthEncoded = 0;
+            }
+
+            return lengthEncoded;
+        }
+
+        public unsafe static int GetBytesAsciiFastPath(Encoding encoding, char* chars, int charCount, byte* bytes, int byteCount)
+        {
+            // Fast path for pure ASCII data for ASCII and UTF8 encoding
+            Debug.Assert(encoding != null);
+            if (bytes == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.bytes, ExceptionResource.ArgumentNull_Array);
+            if (chars == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.chars, ExceptionResource.ArgumentNull_Array);
+            if (charCount < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.charCount, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+            if (byteCount < 0)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.byteCount, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+            Contract.EndContractBlock();
+
+            int lengthEncoded;
+            if (charCount > 0 && byteCount > 0) {
+                lengthEncoded = GetBytesAsciiFastPath(chars, bytes, Math.Min(charCount, byteCount));
+                if (lengthEncoded < byteCount) {
+                    // Not all ASCII, use encoding's GetBytes for remaining conversion
+                    lengthEncoded += encoding.GetBytesFallback(chars + lengthEncoded, charCount - lengthEncoded, bytes + lengthEncoded, byteCount - lengthEncoded, null);
+                }
+            } else {
+                // Nothing to encode
+                lengthEncoded = 0;
+            }
+
+            return lengthEncoded;
+        }
+
+        public unsafe static int GetBytesAsciiFastPath(Encoding encoding, char* chars, int charCount, byte* bytes, int byteCount, EncoderNLS encoder)
+        {
+            // Fast path for pure ASCII data for ASCII and UTF8 encoding
+            // Just need to Assert, this is called by internal EncoderNLS and parameters should already be checked
+            Debug.Assert(encoding != null);
+            Debug.Assert(bytes != null);
+            Debug.Assert(chars != null);
+            Debug.Assert(charCount >= 0);
+            Debug.Assert(byteCount >= 0);
+
+            int lengthEncoded;
+            if ((encoder?.InternalHasFallbackBuffer ?? false) && 
+                (encoder.FallbackBuffer.InternalGetNextChar()) != 0) {
```

Comment:
This should be 
encoder.FallbackBuffer.Remaining > 0

---
In reply to: [96473296](https://github.com/dotnet/coreclr/pull/8969#discussion_r96473296) [](ancestors = 96473296)